### PR TITLE
refactor: wallet 

### PR DESCRIPTION
**Pull request by sorokinvj**
This pr introduces changes for wallet functionality and style:
- new provider that keeps wallet state, and keep track of all etherium operations (network changed, account changed)
- new data to show (Total, available)
- 2 types of wallet: general (available in nav) and booking (estimate before and after booking)
- small fixes in booking flow

### CODE REVIEW

File: pages/login/index.js [hunk 1]
```diff
@@ -2,68 +2,74 @@ import Head from 'next/head';
...
 
-  const walletConnectAndSignInFlow = async () => {
-    setShouldFollowUpOnConnectAndSign(true)
-    if(!isWalletConnected){
-      connectWallet()
-    }else{
-      executeRestOfSignInWithWallet()
+  const walletConnectAndSignInFlow = async (event) => {
+    event.preventDefault();
+    const activated = await injected.activate();
```
Comment (sorokinvj) [inline pages/login/index.js:60]: @juliencharrel this is what `activate` from `useWeb3React` is missing - returning the address that has been activated. However, injected.activate() is doing the job quite well

File: contexts/wallet/index.js [hunk 2]
```diff
@@ -124,26 +185,46 @@ export const useWallet = () => {
...
-  const signMessage = async (msg) => {
+  const signMessage = async (msg, accountId) => {
+    let provider;
+    if (!library) {
+      provider = await injected.getProvider();
+    } else {
+      provider = library.provider;
+    }
     try {
-      const signedMessage = await library.provider.request({
```
Comment (sorokinvj) [inline contexts/wallet/index.js:129]: here, I also use one of the `injected` methods, otherwise as with "Sign In with wallet" the library is not ready when we need it @juliencharrel

File: utils/blockchain.js [hunk 3]
```diff
@@ -55,3 +63,51 @@ export async function sendDAOToken(library, toAddress, amount) {
...
+    return acc.gt(curr) ? acc : curr;
+  }, BigNumber.from(0));
+  const maxStake = formatBigNumberForDisplay(maxStakeBigNum, 18);
+
+  const tokensToStake = Math.max(
+    totalBookingTokenCost - Math.max(maxStake, currentYearStake),
+    0,
+  );
+
+  return tokensToStake;
```
Comment (sorokinvj) [inline utils/blockchain.js:112]: @juliencharrel I rewrote your function, could you review please? First, I tried to make it more readable in case somebody else would need to use it (or even me in 2-3 months), Secondly, bookedDates is already flat in my case (intentionally), so direct copy/paste from your code doesn't work. 

I also want to have a unit test for all the helpers and this one in particular, lets connect for 10-15 min to defined edge cases and just set some examples?

**Review by juliencharrel — CHANGES_REQUESTED** [review]

**Comment by closerearth** [comment]
I get `Application error: a client-side exception has occurred (see the browser console for more information).` when I open preview.

**Comment by sorokinvj** [comment]
> I get `Application error: a client-side exception has occurred (see the browser console for more information).` when I open preview.
> 
> ```
> framework-79bce4a3a540b080.js:1 TypeError: Cannot read properties of null (reading 'walletAddress')```
> ```

yes, sorry, fixed

File: utils/blockchain.js (hunk 3 above)
Comment (sorokinvj) [reply utils/blockchain.js:112]: created a ticket for that https://github.com/closerdao/closer-ui/issues/49
